aiidalab_widgets_base/structures.py

```diff
@@ -136,49 +123,68 @@ def store_structure(self, label=None, description=None):
         self.structure_node.store()
         print("Stored in AiiDA: " + repr(self.structure_node))
 
-    def freeze(self):
-        """Do not allow any further modifications"""
-        self._structure_sources_tab.layout.visibility = 'hidden'
-        self.frozen = True
-        self.btn_store.disabled = True
-        self.structure_description.disabled = True
-        self.data_format.disabled = True
-
-    @property
-    def node_class(self):
-        return self.data_format.value
-
-    @node_class.setter
-    def node_class(self, value):
-        if self.frozen:
-            return
-        self.data_format.value = value
-
-    @property
-    def structure_node(self):
-        """Returns AiiDA StructureData node."""
-        if self._structure_node is None:
-            if self.structure_ase is None:
-                return None
-            # perform conversion
-            if self.data_format.value == 'CifData':
-                from aiida.orm.nodes.data.cif import CifData
-                self._structure_node = CifData()
-                self._structure_node.set_ase(self.structure_ase)
-            else:  # Target format is StructureData
-                self._structure_node = StructureData(ase=self.structure_ase)
-            self._structure_node.description = self.structure_description.value
-            self._structure_node.label = self.structure_ase.get_chemical_formula()
-        return self._structure_node
+    @observe('node_class')
+    def _change_structure_node(self, change):
+        if change['new'] is not None:
+            self.structure = self._inserted_structure
+
+    @validate('structure')
+    def _valid_structure(self, change):
+        """Returns ASE atoms object and sets structure_node trait."""
+
+        self._inserted_structure = change['value']
```

Comment:
This function should not have any side-effects besides transforming the value and returning it.

Reply:
Indeed. When I set the value to `None` it is not even called. So I agree that all 'side-effects' should go into the observed methods.